server/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -612,6 +613,15 @@ public RestApiVersion getRestApiVersion() {
         return restApiVersion;
     }
 
+    public void markResponseRestricted(String restriction) {
+        if (params.containsKey(RESPONSE_RESTRICTED)) {
+            throw new IllegalArgumentException("The parameter [" + RESPONSE_RESTRICTED + "] is already defined.");
+        }
+        params.put(RESPONSE_RESTRICTED, restriction);
+        // this parameter is intended be consumed via ToXContent.Params.param(..), not this.params(..) so don't require it is consumed here
+        consumedParams.add(RESPONSE_RESTRICTED);
```

Comment:
I just realised that we have

https://github.com/elastic/elasticsearch/blob/b3d7ed530aaf83e5a6decfdf391490fc74158d8d/server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java#L186-L188

On the surface, it seems to be the place to define unconsumed response parameters. But it requires a few more refactor to work. So I am OK with actively consuming it here.